uxarray/grid.py

```diff
@@ -0,0 +1,278 @@
+
+    # vertices init
+    def __from_vert__(self):
+        """Create a grid with one face with vertices specified by the given
+        argument."""
+        self.__init_mesh2__()
+        self.in_ds.Mesh2.attrs['topology_dimension'] = self.vertices[0].size
+
+        self.in_ds["Mesh2"].topology_dimension
+
+        x_coord = self.vertices.transpose()[0]
+        y_coord = self.vertices.transpose()[1]
+
+        # single face with all nodes
+        num_nodes = x_coord.size
+        conn = list(range(0, num_nodes))
+        conn = [conn]
+
+        self.in_ds["Mesh2_node_x"] = xr.DataArray(data=xr.DataArray(x_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_node_y"] = xr.DataArray(data=xr.DataArray(y_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_face_nodes"] = xr.DataArray(
+            data=xr.DataArray(conn),
+            dims=["nMesh2_face", "nMaxMesh2_face_nodes"],
+            attrs={
+                "cf_role": "face_node_connectivity",
+                "_FillValue": -1,
+                "start_index": 0
+            })
+
+    # TODO: gridspec init
+    def __from_gridspec__(self):
+        print("initializing with gridspec")
+
+    # load mesh from a file
+    def __from_file__(self):
+        """Loads a mesh file Also, called by __init__ routine This routine will
+        automatically detect if it is a UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+        # call function to set mesh file type: self.mesh_filetype
+        self.__find_type__()
+
+        # call reader as per mesh_filetype
+        if self.mesh_filetype == "exo":
+            self.read_exodus(self.ext_ds)
+        elif self.mesh_filetype == "scrip":
+            self.read_scrip(self.ext_ds)
+        elif self.mesh_filetype == "ugrid":
+            self.read_ugrid(self.filepath)
+        elif self.mesh_filetype == "shp":
+            self.read_shpfile(self.filepath)
+        else:
+            raise RuntimeError("unknown file format:" + self.mesh_filetype)
+
+    # helper function to find file type
+    def __find_type__(self):
+        """Checks file path and contents to determine file type Also, called by
+        __init__ routine This routine will automatically detect if it is a
+        UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+
+        try:
+            # extract the file name and extension
+            path = PurePath(self.filepath)
+            file_extension = path.suffix
+
+            # try to open file with xarray
+            self.ext_ds = xr.open_dataset(self.filepath, mask_and_scale=False)
+            #
+        except (TypeError, AttributeError) as e:
+            msg = str(e) + ': {}'.format(self.filepath)
+            print(msg)
+            raise RuntimeError(msg)
+            exit()
```

Comment:
This set of lines is duplicated three times, and could be moved to a `finally:` block.